### PR TITLE
Add frontend nginx container for logging and plausible proxying

### PR DESCRIPTION
**Pull request by sarayourfriend**
## Fixes

Related to #536 by @sarayourfriend 
Related to #2996 by @obulat
Related to #2993 (does not close because we also need to deploy the nginx container to live environments)

The issue will only be fixed, if we take this approach, after adding the new service to the live deployments.

## Description

Adds a new Nginx image for us to publish and use as a reverse-proxy to the frontend service. This enables two things:

1. Per-request logging in a format identical to the API logging. There are improvements we can make here with respect to configuration sharing, but that adds complexity that we can push off for now while we validate this approach.
2. Plausible request proxying that is testable locally.

This PR is still a draft because I need to add building and publishing the image to CI.

## Testing Instructions

Run `just frontend/up` which will automatically build the new `frontend_nginx` service, then `just frontend/run dev`. Afterwards, to be able to test Plausible locally, run `just frontend/init`. Run `just ps` and visit the `frontend_nginx` URL and confirm you can visit the site and browse it as normal. Audit the nginx logs using `just logs frontend_nginx`. Next, test Plausible by visiting it using the [instructions from the Openverse documentation](https://docs.openverse.org/frontend/guides/analytics.html#access-plausible). Confirm that you see the page views as expected.

To test that upstream proxying works, modify `OPENVERSE_NGINX_PLAUSIBLE_EVENT_URL` to point to upstream (`https://plausible.io/api/event`) and confirm the requests are forwarded by Nginx.

Note that in the Plausible documentation for Nginx proxying they also have you proxy the script. However, we bundle the script into the app via the vue plugin, so this isn't necessary for us to do (as you can see from the original configuration, we were not proxying the script).

_Edit: I (@obulat) added the issue that this PR fixes_ 

## Checklist

- [x] My pull request has a descriptive title (not a vague title like`Update index.md`).
- [x] My pull request targets the _default_ branch of the repository (`main`) or a parent feature branch.
- [x] My commit messages follow [best practices][best_practices].
- [x] My code follows the established code style of the repository.
- [N/A] I added or updated tests for the changes I made (if applicable).
- [x] I added or updated documentation (if applicable).
- [x] I tried running the project locally and verified that there are no visible errors.
- [N/A] I ran the DAG documentation generator (if applicable).

[best_practices]:
  https://git-scm.com/book/en/v2/Distributed-Git-Contributing-to-a-Project#_commit_guidelines

## Developer Certificate of Origin

<details>
<summary>Developer Certificate of Origin</summary>

```
Developer Certificate of Origin
Version 1.1

Copyright (C) 2004, 2006 The Linux Foundation and its contributors.
1 Letterman Drive
Suite D4700
San Francisco, CA, 94129

Everyone is permitted to copy and distribute verbatim copies of this
license document, but changing it is not allowed.


Developer's Certificate of Origin 1.1

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I
    have the right to submit it under the open source license
    indicated in the file; or

(b) The contribution is based upon previous work that, to the best
    of my knowledge, is covered under an appropriate open source
    license and I have the right under that license to submit that
    work with modifications, whether created in whole or in part
    by me, under the same open source license (unless I am
    permitted to submit under a different license), as indicated
    in the file; or

(c) The contribution was provided directly to me by some other
    person who certified (a), (b) or (c) and I have not modified
    it.

(d) I understand and agree that this project and the contribution
    are public and that a record of the contribution (including all
    personal information I submit with it, including my sign-off) is
    maintained indefinitely and may be redistributed consistent with
    this project or the open source license(s) involved.
```

</details>

### CODE REVIEW

**Review by obulat — COMMENTED** [review]
When I run `just frontend/up` and `just frontend/dev`, I can only see the analytics requests logged locally. Is there a way to see the normal frontend access logs, too?

**Comment by sarayourfriend** [comment]
> Is there a way to see the normal frontend access logs

What do you mean by "the normal frontend access logs"? The Nginx container is the "logging" entrypoint, in production it would front Nuxt. You can view those logs using `just logs frontend_nginx` as mentioned in the testing instructions.

**Comment by obulat** [comment]
> > Is there a way to see the normal frontend access logs
> 
> Can you clarify what you mean by "the normal frontend access logs"? The Nginx container is the "logging" entrypoint, in production it would front Nuxt. You can view those logs using `just logs frontend_nginx` as mentioned in the testing instructions.

When I tested this PR, I only saw the POST requests sent for analytics. I did not see the logs for requests of the frontend pages when I visited them. So, nothing for a GET request to "localhost:8443" or a server-rendered "localhost:8443/image/?q=cat", for instance. Are those requests supposed to be logged in the `frontend_nginx` logs?

**Comment by sarayourfriend** [comment]
> Are those requests supposed to be logged in the frontend_nginx logs?

Yes, the idea is that in production the Nginx container would proxy the Nuxt container and provide us with access logs (and Plausible proxying). To test the logs in this PR you have to access the frontend through the `frontend_nginx` service, not directly at `localhost:8443`, and tail the logs for `frontend_nginx`.

I'm realising the testing instructions aren't clear about this, it just says "visit the `frontend_nginx` URL", which isn't obvious. You should test it by accessing the frontend at `localhost:50290` and tailing the nginx logs. You need to have the frontend running locally though, otherwise the Nginx service won't have anything to send the requests to.

**Comment by obulat** [comment]
Thank you for clarifications, @sarayourfriend ! 
Is there a way to remove duplicated logging:
```

192.168.214.1 - - [06/Oct/2023:07:23:42 +0000] "GET /_nuxt/pages/index.modern.js HTTP/1.1" 200 6006 "http://0.0.0.0:50290/" "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/117.0.0.0 Safari/537.36" "-"
{"time_local":"06/Oct/2023:07:23:42 +0000","remote_addr":"x.x.x.x","remote_user":"","request":"GET /_nuxt/pages/index.modern.js HTTP/1.1","status": "200","host_header": "0.0.0.0","body_bytes_sent":6006,"request_time":"0.006","http_referrer":"http://0.0.0.0:50290/","http_user_agent":"Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/117.0.0.0 Safari/537.36","upstream_response_time":0.006,"http_x_forwarded_for":""}
192.168.214.1 - - [06/Oct/2023:07:23:42 +0000] "GET /_nuxt/9c7bd6e0532911068731cb497475a9a5.svg HTTP/1.1" 200 11886 "http://0.0.0.0:50290/" "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/117.0.0.0 Safari/537.36" "-"
{"time_local":"06/Oct/2023:07:23:42 +0000","remote_addr":"x.x.x.x","remote_user":"","request":"GET /_nuxt/9c7bd6e0532911068731cb497475a9a5.svg HTTP/1.1","status": "200","host_header": "0.0.0.0","body_bytes_sent":11886,"request_time":"0.010","http_referrer":"http://0.0.0.0:50290/","http_user_agent":"Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/117.0.0.0 Safari/537.36","upstream_response_time":0.010,"http_x_forwarded_for":""}
```

Also, it seems that these logs will be pretty noisy if we log every request for every JS file...

**Review by AetherUnbound — COMMENTED** [review]
The implementation looks good! +1 to all the points @obulat brings up, it would be good to emit just the JSON logs and to remove all of the `_nuxt/` access calls as well.

**Comment by sarayourfriend** [comment]
Those are great changes, Olga. We should apply them to the API configuration as well. I guess it would just about cut the size of our Nginx logs by at least a third, if not closer to a half! Talk about a real cost saver, especially as we're adding new logging for the frontend.

Filtering the `_nuxt` requests out could be good too, I hadn't considered how noisy they are; however, those are also requests that Nuxt is needing to handle, having visibility into them could be important (maybe they're particularly slow, sometimes, for example). Theoretically those resources are cached in production, and we wouldn't see more than a single log per-resource per-release, provided our caching works the way we think it should. And if it doesn't, by chance, what a great way to find out, by seeing more logs here than we expect! Do you think it's worth keeping them? If we see them clogging up production request logs then it would be a great signal for us to check our caching configuration :slightly_smiling_face:

**Comment by obulat** [comment]
> Filtering the `_nuxt` requests out could be good too, I hadn't considered how noisy they are; however, those are also requests that Nuxt is needing to handle, having visibility into them could be important (maybe they're particularly slow, sometimes, for example). Theoretically those resources are cached in production, and we wouldn't see more than a single log per-resource per-release, provided our caching works the way we think it should. And if it doesn't, by chance, what a great way to find out, by seeing more logs here than we expect! Do you think it's worth keeping them? If we see them clogging up production request logs then it would be a great signal for us to check our caching configuration 🙂

I hadn't considered how the `_nuxt/` requests are cached, and yes, I think it's best to keep those. If they show that the caching configuration isn't set up correctly, then we can fix the caching and they will stop clogging the logs :)

File: frontend/nginx.conf.template [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+gzip_vary on;
+gzip_proxied expired private auth;
+gzip_types application/json text/plain application/javascript;
+gzip_disable "MSIE [1-6]\.";
+
+upstream ov_service {
+    server          $OPENVERSE_NGINX_UPSTREAM_URL;
+}
+
+server {
```
Comment (sarayourfriend) [inline frontend/nginx.conf.template:45]:
```suggestion
server {
    access_log  /var/log/nginx/access.log  json_combined;

```

**Review by obulat — APPROVED** [review]
LGTM!

**Review by AetherUnbound — APPROVED** [review]
Fantastic!

File: frontend/Dockerfile.nginx [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+# contexts required for the frontend build (namely, `repo_root`). These are
+# possible with compose as of https://github.com/docker/compose/pull/10369
+# and do appear to work, but it increases the complexity and these features
+# are not yet all available in every containerd implementation or distribution
+# (Debian, for example, still distributes an older version of compose without
+# this new feature). For the sake of keeping the local setup simpler and not
+# requiring contributors to have bleeding-edge Docker and compose plugin versions,
+# we can simply use a separate file for the time being. Once/if we start to use
+# nginx to serve static files for the frontend, we will need to integrate this
+# target into the shared Dockerfile.
```
[the remark this answers is not in the document]
Comment (sarayourfriend) [reply frontend/Dockerfile.nginx:11]: Yes :+1: Done here https://github.com/WordPress/openverse/issues/3180

**Pull request merged**